src/derivative.jl

```diff
@@ -15,14 +15,16 @@ This method assumes that `isa(f(x), Union{Real,AbstractArray})`.
 end
 
 """
-    ForwardDiff.derivative(f!, y::AbstractArray, x::Real, cfg::DerivativeConfig = DerivativeConfig(f!, y, x))
+    ForwardDiff.derivative(f!, y::AbstractArray, x::Real, cfg::DerivativeConfig = DerivativeConfig(f!, y, x), check=Val{true}())
```

Comment:
Instead of doing this, can't we just disable checking by dispatching on the `Void` tag? That would be cleaner and less of an API change.

Reply:
That's what I was referring to here: https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/200#issuecomment-340879224 in that now a user cannot turn off the tag checking in a call to another package if you can't send that extra argument.

Reply:
Short answer: no, and we should generally discourage use of the `Void` tag, especially in packages.

The problem is that it reopens the door to perturbation confusion. For example, [its use in NLsolve.jl](https://github.com/JuliaNLSolvers/NLsolve.jl/blob/535b7361c700b39b6f1f1148f805c6036491ef55/src/autodiff.jl#L7) means that if the function passed to it _also_ uses `Void` tags then there is potential for confusion.

A better pattern is for the library to generate its own tag, and use that, similar to what I did for Hessian:
https://github.com/JuliaDiff/ForwardDiff.jl/pull/279/files#diff-7957397746d48056ebb4bb3af5bab442R14

Reply:
The problem with the previous change though was that generating your own tag no longer works, which is what I was doing before.